mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -5,821 +5,745 @@
     #[snafu(display(
         "Failed to allocate buffer for reading {}: {}",
         name,
         source
     ))]
-    ReadAlloc { name: String, source: DmaError },
+    ReadAlloc { source: DmaError, name: String },
     #[snafu(display(
         "Failed to allocate buffer for writing {}: {}",
         name,
         source
     ))]
-    WriteAlloc { name: String, source: DmaError },
-    #[snafu(display("Serialization error: {}", source))]
-    SerializeError { source: Error },
-    #[snafu(display("Deserialization error: {}", source))]
-    DeserializeError { source: Error },
+    WriteAlloc { source: DmaError, name: String },
+    #[snafu(display("Error reading {}: {}", name, source))]
+    ReadError { source: CoreError, name: String },
+    #[snafu(display("Error writing {}: {}", name, source))]
+    WriteError { source: CoreError, name: String },
     #[snafu(display(
-        "Incorrect MetaData header size: actual={} expected={}",
-        actual_size,
-        expected_size
+        "Failed to obtain BdevHandle for child {}: {}",
+        name,
+        source
     ))]
-    HeaderSize {
-        actual_size: u32,
-        expected_size: u32,
-    },
-    #[snafu(display("Incorrect MetaData header signature"))]
-    HeaderSignature {},
-    #[snafu(display("Incorrect MetaData header checksum"))]
-    HeaderChecksum {},
-    #[snafu(display("Incorrect MetaData index checksum"))]
-    IndexChecksum {},
-    #[snafu(display("Incorrect MetaData configuration object checksum"))]
+    HandleError { source: CoreError, name: String },
+    #[snafu(display("Incorrect MetaDataObject signature"))]
+    ObjectSignature {},
+    #[snafu(display("Incorrect MetaDataObject checksum"))]
     ObjectChecksum {},
-    #[snafu(display("MetaData index is inconsistent"))]
-    IndexInconsistent {},
-    #[snafu(display("MetaData index ({}) out of range ({})", selected, used))]
-    IndexOutOfRange { selected: u32, used: u32 },
-    #[snafu(display(
-        "Number of objects ({}) exceeds index size ({})",
-        used,
-        size
-    ))]
-    IndexSizeExceeded { used: u32, size: u32 },
-    #[snafu(display("Not enough space left on MetaData partition"))]
-    PartitionSizeExceeded {},
-    #[snafu(display("MetaData partition is missing or invalid"))]
-    MissingPartition {},
-    #[snafu(display("Error calculating timestamp: {}", source))]
-    TimeStampError { source: SystemTimeError },
+    #[snafu(display("Incorrect child table checksum"))]
+    ChildTableChecksum {},
+    #[snafu(display("Maximum number of children exceeded"))]
+    ChildTableSize {},
+    #[snafu(display("Incorrect MetaDataIndex signature"))]
+    IndexSignature {},
+    #[snafu(display("Incorrect MetaDataIndex size"))]
+    IndexSize {},
+    #[snafu(display("Incorrect MetaDataIndex checksum"))]
+    IndexChecksum {},
+}
+
+#[derive(Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct MetaDataChildEntry {
+    pub guid: Guid,
+    pub state: u16,
```

Comment:
just now we store only the state of the child, so we can tell which replica was online and which was degraded or even faulted. I'm not sure if generation number isn't a more valuable information when it comes to the rebuild. I wasn't able to imagine a case when we could not detect wrong replica (after nexus is recreated) however increasing generation number is simpler to do, does not rely on representation of state by a number that can potentially change. What do you think Colin and @gila ? Some time ago we discussed RPC method for obtaining the generation number that would be used by the operator to decide what to do when importing a volume. So that sounds like a more useful information.

Reply:
Each "index" entry represents a snapshot of the nexus, and contains a timestap, generation number, and the current state of each child (identified by it's uuid).  So each child in effect inherits the same generation number. The idea is that a new index entry is created whenever the status of any child changes. Also, the intention is that the state of each child is represented as a bitmap - where the meaning of each bit is yet to be determined. However the code here at the persistence layer does not need to know what the "state" represents - it's just a value that needs to be stored and retrieved.

And, yes there will be gRPC methods to retrieve the latest index entry from one or all children.

Reply:
You are right. The generation number in the index is what I was looking for. Looks good 👍